File: jar-common/src/main/java/com/beancounter/common/model/CurrencyKeyDeserializer.java
package com.beancounter.common.model;

import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.KeyDeserializer;

public class CurrencyKeyDeserializer extends KeyDeserializer {
  @Override
  public CurrencyPair deserializeKey(String key, DeserializationContext ctxt) {
    return CurrencyPair.builder()
        .from(key.substring(0, 3))
        .to(key.substring(4, 7)).build();
  }
}

File: jar-contracts/src/main/kotlin/com/beancounter/common/utils/CurrencyKeyDeserializer.kt
package com.beancounter.common.utils

import com.beancounter.common.model.IsoCurrencyPair
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.KeyDeserializer

/**
 * Convert the String into an Object representation. Jackson
 */
class CurrencyKeyDeserializer : KeyDeserializer() {
    override fun deserializeKey(
        key: String,
        ctxt: DeserializationContext
    ): IsoCurrencyPair =
        IsoCurrencyPair(
            key.substring(
                0,
                3
            ),
            key.substring(
                4,
                7
            )
        )
}